### PR TITLE
Daml: Add basic support for BigNumeric in 1.dev

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This is part of #8719 

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-stdlib-src/DA/BigNumeric.daml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+roundingFloor = primitive @"BERoundingFloor"
+roundingHalfUp = primitive @"BERoundingHalfUp"
+roundingHalfDown = primitive @"BERoundingHalfDown"
+roundingHalfEven = primitive @"BERoundingHalfEven"
+roundingUnnecessary = primitive @"BERoundingUnnecessary"
+
+zero, one, ten: BigNumeric
+zero = fromNumeric (0.0: Numeric 0)
+one = fromNumeric (1.0: Numeric 0)
+ten = fromNumeric (10.0: Numeric 0)
```
Comment (cocreature) [inline compiler/damlc/daml-stdlib-src/DA/BigNumeric.daml:28]: What is this for?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala [hunk 2]
```diff
@@ -360,7 +361,7 @@ private[lf] object Pretty {
   def prettyValue(verbose: Boolean)(v: Value[ContractId]): Doc =
     v match {
       case ValueInt64(i) => str(i)
-      case ValueNumeric(d) => str(d)
+      case ValueNumeric(d) => text(data.Numeric.toString(d))
```
Comment (cocreature) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala:364]: does this change anything? I wasn’t expecting to see a change to `Numeric` in here. I guess the pretty printer doesn’t matter much so no issues with keeping it if it changes it in a useful way.

**Review by cocreature — APPROVED** [review]
lgtm I think there is a bit of cleanup left in `DA.BigNumeric` but you already added a TODO for that and I have no issues with addressing that in a follow-up PR.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Pretty.scala:364]: I just realized the Numeric are printed in a consistent way with LF builtin to TO_TEXT_NUMERIC.
This is used in the output of LF Repl wish it is match with EXPETED.ledger.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs [hunk 3]
```diff
@@ -350,6 +350,17 @@ data BuiltinExpr
...
+  | BEScaleBigNumeric              -- :: BigNumeric -> Int64
+  | BEPrecisionBigNumeric              -- :: BigNumeric -> Int64
+  | BEAddBigNumeric              -- :: BigNumeric -> BigNumeric -> BigNumeric
+  | BESubBigNumeric              -- :: BigNumeric -> BigNumeric -> BigNumeric
+  | BEMulBigNumeric              -- :: BigNumeric -> BigNumeric -> BigNumeric
+  | BEDivBigNumeric              -- :: Int64 -> RoundingMode -> BigNumeric -> BigNumeric -> BigNumeric
+  | BEShiftBigNumeric            -- :: Int64 -> BigNumeric -> BigNumeric
+  | BEToNumericBigNumeric        -- :: ∀(s:nat). BigNumeric -> Numeric s
+  | BEFromNumericBigNumeric        -- :: ∀(s:nat). Numeric s -> BigNumeric
+
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Base.hs:363]: fixed.

File: compiler/damlc/daml-prim-src/GHC/Types.daml [hunk 4]
```diff
@@ -197,4 +197,4 @@ data Decimal =
   Decimal Opaque
 
 
-#endif
+#endif
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply compiler/damlc/daml-prim-src/GHC/Types.daml:200]: fixed.

File: compiler/damlc/daml-stdlib-src/DA/BigNumeric.daml (hunk 1 above)
Comment (remyhaemmerle-da) [reply compiler/damlc/daml-stdlib-src/DA/BigNumeric.daml:28]: It is just useful constants. For instance,  java.math.BigDecimal defines ZERO, ONE, TEN.
Whoever will clean this file is free to dropped them.

**Review by remyhaemmerle-da — COMMENTED** [review]
@cocreature I will let you or @sofiafaro-da polish the integration in Daml (in particular the content of DA.BigNumeric). This is the minimal I need to start playing with BigNumeric and run some benchmarks.

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs [hunk 5]
```diff
@@ -125,6 +125,15 @@ safetyStep = \case
       BERoundNumeric      -> Safe 1
       BECastNumeric       -> Safe 0
       BEShiftNumeric      -> Safe 1
+      BEScaleBigNumeric     -> Safe 1
```
Comment (remyhaemmerle-da) [inline compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:128]:
```suggestion
      -- TODO https://github.com/digital-asset/daml/issues/8719
      --  review carfully BitNumeric functions
      BEScaleBigNumeric     -> Safe 1
```

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs [hunk 6]
```diff
@@ -125,6 +125,15 @@ safetyStep = \case
...
+      BEPrecisionBigNumeric -> Safe 1
+      BEAddBigNumeric       -> Safe 1
+      BESubBigNumeric       -> Safe 1
+      BEMulBigNumeric       -> Safe 1
+      BEDivBigNumeric       -> Safe 1
+      BEShiftBigNumeric     -> Safe 1
+      BEToNumericBigNumeric -> Safe 0
+      BEFromNumericBigNumeric -> Safe 0
```
Comment (remyhaemmerle-da) [inline compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:138]: @sofiafaro-da or @moritzkiefer-da Could you review that when you cleaning up the compiler?

Comment (cocreature) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:138]: can you add a todo please

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs (hunk 5 above)
Comment (remyhaemmerle-da) [inline compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:130]:
```suggestion
      -- TODO https://github.com/digital-asset/daml/issues/8719
      --  review carefully BitNumeric functions
      BEScaleBigNumeric     -> Safe 1
```

File: compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs (hunk 6 above)
Comment (remyhaemmerle-da) [reply compiler/daml-lf-tools/src/DA/Daml/LF/Simplifier.hs:138]: done.

**Pull request merged**